### ISSUE TITLE
In pypi, it is impossible to reupload a removed file.

### ISSUE THREAD

**Issue by Natim** [pypa/packaging-problems#74]
HTTPError: 400 Client Error: This filename has previously been used, you should use a different version.

**Comment by Natim** [comment]
Also the previous version has been removed and is impossible to find.

**Comment by daenney** [comment]
It's probably still available in the Fastly caches, which is why you need to use a new filename. The old filename will have been marked as to cache indefinitely so even if you could upload a filename with the same name, once it got served to a client they would never get the new version you uploaded.

**Comment by Natim** [comment]
I nmy case it isn't a problem because it is the exact same file.

**Comment by hickford** [comment]
See Donald's email at http://comments.gmane.org/gmane.comp.python.distutils.devel/22739

>  I've pushed changes to PyPI where it is no longer possible to reuse a filename and attempting to do it will give an 400 error "This filename has previously been used, you should use a different version."

This is an FAQ, I think it should be documented officially somewhere.

**Comment by hickford** [comment]
Npm did the same in 2014. See http://blog.npmjs.org/post/77758351673/no-more-npm-publish-f

> While it is annoying to have to bump the version number for typos documentation changes, **I believe in the long run, the benefits of greater reliability and data integrity are well worth it.**

I presume the justification is the same for PyPI. It's an FAQ, so should probably go in documentation somewhere.

**Comment by Natim** [comment]
Then we shouldn't allow people to remove their files if they cannot put them back.

**Comment by Natim** [comment]
I think we should allow to reupload the same removed file

**Comment by tylerdave** [comment]
There are very good reasons for the current behavior. Authors should be able to delete for any number of reasons (legal, security, etc.) Users of the package should be able to rely on getting the exact same thing every time they install a package of a specific version.

If you delete a package that someone relies on, they know the version is gone and they need to make a change to fix it. If you could delete a package and replace it with something different but with the same version, it can break their program is any number of subtle ways and it would be very hard to determine the cause of the problem.

Allowing this would break the entire version number contract. You may have what seems to be a good reason to replace a version but allowing it is not worth making versions unreliable.

**Comment by hickford** [comment]
Absolutely.

**Comment by Natim** [comment]
> If you delete a package that someone relies on

You broke their package and you cannot put it back.

**Comment by Natim** [comment]
> If you could delete a package and replace it with something different but with the same version, it can break their program

That's not what I am asking for.

I am asking for putting back the package I removed.

**Comment by Natim** [comment]
> Allowing this would break the entire version number contract.

Allowing to put back the version you removed doesn't break any contracts. + You already have the previous package hash so you can check the version didn't change.

**Comment by tylerdave** [comment]
There I agree. If it can be ensured via the hash that *only* the *exact same* package is uploaded to the same version then I don't see this being a problem in concept.

**Comment by hickford** [comment]
If the documentation and confirmation makes it clear that unpublishing is permanent, then I think it's reasonable and prudent. 

> It is generally considered bad behavior to remove versions of a library that others are depending on! Even if a package version is unpublished, that specific name and version combination can never be reused. In order to publish the package again, a new version number must be used.

https://docs.npmjs.com/cli/unpublish

**Comment by hickford** [comment]
Could @Natim's 'same hash' example be abused maliciously? Upload malicious version, immediately delete it. Upload nice version. Wait until you have thousands of users. Delete nice version, reupload nasty version (same hash).

I think the policy should be strengthened to 'no uploads to old versions' https://github.com/pypa/packaging-problems/issues/75

**Comment by dstufft** [comment]
Unless you have the physical file laying around still, it's unlikely you're going to have something that matches the same hash. The ``setup.py sdist`` command does not have deterministic output, each time you run it even if the code hasn't changed. This also means you can't use ``setup.py`` to upload the file, since ``setup.py`` will only let you upload a file that it has created in the currently executing command, not an already created file. That doesn't make it impossible to upload a file with the same hash, but it makes it tricky which suggests it's a bad UX to expect authors to have to navigate.

Most likely the eventual solution to this is that "delete" won't actually be a full out absolute deletion, it'll be more like a soft delete where it just acts as if it's deleted without actually deleting it (so it won't show up in the API, won't appear anywhere, etc) but there will be a list of these deleted things when the author logs in and a button that says "Restore" that allows them to restore a file they've previously deleted. Possibly this would have a periodic cleanup where if something was soft deleted for some period of time (a month? 6 months? a year?) we'll go through and clean it up and actually hard delete it then. Perhaps we'd also enable it for authors to trigger an immediate hard delete of something they've soft deleted, but there would be plenty of big warnings that if they press that button there is no recovery possible.

**Comment by Natim** [comment]
> That doesn't make it impossible to upload a file with the same hash, but it makes it tricky which suggests it's a bad UX to expect authors to have to navigate.

With ``twine`` it is as simple as:

    twine upload cliquet-2.5.0-py2.py3-none.whl

**Comment by dstufft** [comment]
Right, I wrote ``twine``, but not everyone uses that so you have to explain to them that they have to use twine to be able to reupload not ``setup.py upload``. In addition you have to explain to them they need the exact same file, not one created the same way. It's fiddly and people will get confused.

**Comment by Natim** [comment]
People are not dumb, if they need to do something complicated they will eventually succeed. The fact is even if they know all the things, they won't be able to do it.

But yeah #75 is a workaround for know, (using .zip instead of .tar.gz for instance)

**Comment by domibarton** [comment]
As I already wrote it in #75 

I think that behaviour is quite OK for the live repo.

Though, to be honest, it's a huge PITA for the test repo. I support integrity and all that stuff on "production" systems. However, developers need to have their code / packages checked somewhere and it's a PITA if you can't upload them same version twice while testing a new release.

There's no other way than the test repo to test your package. With git (or any other SCM) you can easily create a new branch and test it until you're sure everything works. Or if you've a look at PHP Packagist (compose) there's a -dev version for each development branch. On Docker the same, you can test your feature/release branches before tagging and going "live".

With the new policy you basically say: You've ONE SINGLE TRY and that one SHOULD WORK. No chance for a 2nd try. IMHO this isn't the purpose of a testing system and breaks the whole "we've a testing repo" idea. To be honest, I think this only leads to annoyed developers and a lot of "crippled versions" because developers couldn't properly test their versions before going live.

tl;dr:
I suppose you do that on the live system but not on the test system.

**Comment by brianmay** [comment]
In my case, I forgot to sign the upload. It appears once you have uploaded the package it is impossible to fix any problems you made with the upload without making a new release. Even if you just want to upload the exact same version again.

**Comment by daenney** [comment]
But how do you know it is "the exact same version"? Unless it checks the uploads are binary identical it would allow you to upload a totally different release with the same version which can cause any amount of problems.

**Comment by torarnv** [comment]
Just his what @domibarton is describing. What's the point of a test repo if you can't make mistakes?

**Comment by Natim** [comment]
> Just his what @domibarton is describing. What's the point of a test repo if you can't make mistakes?

Why cannot you do ``package x.y.z.dev0`` and then ``package x.y.z.dev1``?

**Comment by torarnv** [comment]
I could, and then having to remember to wipe those temp changes from my working tree before pushing to the live pypi repo.

**Comment by snare** [comment]
This is ridiculous. I understand the motivation, but when the server 500s during an upload and you have a partial file that you can't delete and replace, that's a problem with the system.

**Comment by snare** [comment]
I uploaded a new version of [Voltron](https://pypi.python.org/pypi/voltron) yesterday and the server threw a 500 error during the upload. This resulted in a partial file being hosted as the current wheel for this package. The file size was smaller than my local one, and the hash differed.

This operation needs to be atomic. If the upload fails, you have no opportunity to try again. The only option is to use a different version number, which is not an appropriate solution.

**Comment by Natim** [comment]
Yes I have the same problem with my last uploaded packages.

**Comment by niedakh** [comment]
My files were uploaded broken, due to a connection error, why can't I replace them?!

**Comment by pstch** [comment]
I also suffer from this issue because of PyPI's recurrent HTTP 500 errors, which leads to incomplete uploads : the client errors out, but the file is created, which then makes it impossible to handle the intermittent upload error by simply retrying.

Having to bump version numbers just because of connection errors is quite annoying. Maybe the upload operation could be made atomic, so that it does not create a file unless everything completes normally (possibly by waiting for user confirmation).

**Comment by dstufft** [comment]
This might be helpful: https://mail.python.org/pipermail/distutils-sig/2016-June/029083.html

**Comment by brechtm** [comment]
Even after removing the project on TestPyPI, it is not possible to upload the same version.

**Comment by bittner** [comment]
Will this behavior stay the same forever? It's a bit painful.

I had a syntax error in the long description (so the reStructuredText didn't get converted), and instead of fixing it manually on PyPI I deleted the complete project and started from scratch. Should be the cleanest approach for a new package, shouldn't it?

Instead, now I look like an idiot: There is a [version 0.1.0](https://pypi.python.org/pypi/djangocms-maps/0.1.0) of my package in the index, but I can't upload an installation package for it ever again.  :confused: 

BTW, here is the same question/problem/answer [on StackOverflow](http://stackoverflow.com/questions/28708705/pypi-400-upload-error/28874414).

**Comment by Natim** [comment]
@bittner My bet is that you should probably consider this to be like that for ever and find another solution yes. This policy simplifies a lot the package caching and replication.

For the long description you can change it in the pypi admin. It doesn't fix it in the package but at least it fixes your problem.

**Comment by goanpeca** [comment]
This is silly, I made a mistake in the upload and now I cant correct a mistake?!!?!

**Comment by Natim** [comment]
You can do a new patch release to correct it. The impact is fairly low.

**Comment by goanpeca** [comment]
>You can do a new patch release to correct it.

I uploaded the dev version (0.1.2dev0) cause I forgot to checkut the tag I was uploading (0.1.1)

But now I have to make a new release in Pypi and conda (yes it uses conda) just because of this restriction? yes it makes no sense...

**Comment by goanpeca** [comment]
Furthermore this is the first release, no one is using it and I was able to delete the files to fix the problem, but now I cannot upload

**Comment by Natim** [comment]
You can probably release 0.1.2 cannot you?

**Comment by Natim** [comment]
This limitation makes it a lot easier to handle the CDN caching policy.

**Comment by goanpeca** [comment]
Sure, I can do as many releases as I want, but that still does not change the fact this is a pretty annoying behavior :-(

**Comment by ncoghlan** [comment]
Publishers (or an attacker that has compromised their upload system) silently replacing previously audited software with compromised software is a security vulnerability previously affecting large parts of the internet.

devpi is relatively easy to run locally and allows folks to test their releases before publishing them officially.

**Comment by dstufft** [comment]
It's a small restriction that makes a lot of things a lot more simple. Not only for end users who don't want the meaning of "foobar 1.0" to suddenly change, but also for the tooling itself to be able to make assumptions that simplify the implementation. It *can* be annoying for authors, but luckily it's also something that only rarely is an issue and for which the work around (issue a new version number) is not particularly onerous.

**Comment by atagar** [comment]
If this isn't gonna be fixed then **please** have the 'remove' button issue a warning that the file cannot be re-uploaded. Like brianmay, I made a release via twine that didn't have a sig. The site doesn't allow you to upload one so I removed the file and tried to re-upload it with the exact same tarball and the signature.

Authors expect this to work. Making removal easy but then making it impossible to re-upload is simply awful.

**Comment by atagar** [comment]
Going with a version bump because I cannot continue to block my release on this. You cannot legitimately claim this isn't a bug when it's this easily leaves packages in an unfixable state. This isn't to say you need to allow re-use of file names, but there's multiple issues with the site that make this simply broken...

* No warning when deleting a file that the filename can never be re-uploaded.
* Impossible to upload a sig for an existing release. There's no reason regarding the CDN the site cannot do this. It's simply a bug.
* There doesn't seem to be a way of safely do a dry-run creation of your release. This is the real problem - it's too easy to permanently break your release. I'm now afraid to run twine commands because I have no clue if it'll do what I want until it either works or permanently breaks my package.

I should be able to create a release, confirm it looks right in PyPI, then publish it. As it is I now need to create a scrap PyPI project so I can safely try twine commands without screwing up my releases.

I sincerely hope the new PyPI site behaves better.

**Comment by mikofski** [comment]
@atagar see [my comment above](https://github.com/pypa/packaging-problems/issues/74#issuecomment-232592817)
>Just use the manual interface by going to your package pypi site and select files. Change the names of the wheels slightly and add a comment if desired. Then you can upload the replacements for the removed files. This worked fine for me.

EG: if your file was `newpackage-0.X-py2-none-any.whl` then change it to `newpackage-0.X-CORRECTED-py2-none-any.whl` and then upload it manually. I did this for [SolarUtils-0.2.2](https://pypi.python.org/pypi/SolarUtils), I added an "a" and it worked fine.

Hope this works for you too!

**Comment by atagar** [comment]
In the past I did that too, creating a 'Stem-1.4.2b.tar.gz' file to work around this. Couple folks that package my project for distributions (don't recall if it was Debian, Gentoo, Arch, or BSD) contacted me to say this created problems for them.

**Comment by mikofski** [comment]
I like your idea of a warning button, or maybe an alert pop up on the remove button to confirm and warn about this feature/issue. That should be part of [Warehouse](https://github.com/pypa/warehouse) - maybe make an issue there?

Also the idea of an optional dry run could also be a Warehouse feature. IE:

    $ python setup.py upload sdist --dry-run

then when you go to [Warehouse](https://pypi.org/) your release would not yet be visible, but would require a confirmation before becoming permanent. Once permanent then it removing it would make it not possible to upload another release of the same version, but the remove button would provide a warning of this first.

**Comment by merwok** [comment]
There is TestPyPI if you want to do dry-run releases and aren’t running a local devpi server: https://wiki.python.org/moin/TestPyPI

**Comment by jianli** [comment]
It seems that the current restrictions on re-uploading the exact same version (as onerous as they are) do not actually prevent a malicious user from uploading multiple files with semantically equivalent versions. The malicious user can upload distinct packages for each of

```python
python setup.py sdist upload 1.0
python setup.py sdist upload 1.0.0
python setup.py sdist upload 1.0.0.0
```

Then, the end-user who has previously verified `1.0` and runs `pip install 1.0` would potentially get an unverified file?

**Comment by brechtm** [comment]
@merwok IIRC TestPyPI also doesn't allow overwriting files either, so if you screw up you can't test with the same version number.

**Comment by brechtm** [comment]
+1 for @mikofski's suggestion for a sort of _staging area_ on PyPI, similar to the functionality that devpi offers. This would allow you to upload a release, test it by supplying a different server to pip's `--index-url` option, and then "commit" the release.

But the question is, who will implement this? ...

**Comment by dstufft** [comment]
See pypa/warehouse#726.

**Comment by ncoghlan** [comment]
I'll also note that if anyone in this thread is using Python for professional purposes and would like to see additional development effort put towards PyPI, that the PSF accepts targeted contributions towards that end: https://donate.pypi.io/

It may also be worth contacting your Python support vendor or deployment platform provider, and asking them to invest time directly in upstream PyPI maintenance and development (since providing a Python runtime is only the first step in effectively supporting Python developers - vendors and platform providers should also be expected to provide reliable access to the wider Python ecosystem).

**Comment by atagar** [comment]
There we go. Took a handful of dry-run attempts but finally got a release I'm happy with. In case others want a safe approach for making releases here's what we're doing...

* [Instructions](https://trac.torproject.org/projects/tor/wiki/doc/stem#ReleaseNotes)
* [Commit](https://gitweb.torproject.org/stem.git/commit/?id=bf34c2bf5fdb6eea84d9589225783ec5b16e8278)

Problems-between-chair-and-keyboard aside here's a couple things that accounted for some of the trial-and-error attempts...

* A multi-line 'description' in the setup.py successfully makes a package and uploads. PyPI however quietly crops the PKG-INFO, causing any data following it (like Categories) to be omitted.
* Evidently you can't [set the bug tracker url through the setup.py](https://stackoverflow.com/questions/14459828/how-to-set-bug-tracker-url-in-setup-py-script).

C'est la vie. Now that we have a safe procedure for making releases I'm happy.

**Comment by merwok** [comment]
Two notes:

1) The `provides` field is not useful and not used, it would be best to remove it to avoid more confusion.

2) You can use setup.py upload for HTTPS uploads, it depends on your configuration in ~/.pypirc

**Comment by atagar** [comment]
Thanks! [Dropped the provides clause](https://gitweb.torproject.org/stem.git/commit/?id=7dbcf307a3f6abd92c1914ee6026664644c67c41). As for setup.py maybe PyPI could provide a more useful error than "Upload failed (403): Must access using HTTPS instead of HTTP"? If it gave a link to the [upload instructions](https://packaging.python.org/distributing/#upload-your-distributions) **and** that told people they can make a pypirc that would be useful. Unfortunately a quick search didn't have that.

**Comment by joernhees** [comment]
this is quite annoying... i just had a wheel for py2 pick up a build/src for py3 (that went through 2to3)... resulting `...-py2-none-any.whl` is broken.

Considering that i noticed the problem within 2 minutes and immediately deleted the bogus `.whl`, it would be nice to have an option to pretend it just didn't happen. I see the arguments that releases should be immutable, but i'm quite sure that no one besides me pulled that file within the time-frame. As mistakes can and will happen, i'd love to not be robbed of a choice as a maintainer to decide what's best myself.

Why not at least have a grace period of a few minutes?

**Comment by dstufft** [comment]
2 minutes is plenty of time for hundreds of mirroring clients that run on a one minute cron job to find the file and pull it down immediately.

**Comment by joernhees** [comment]
mirroring clients as in pypi caches? they could and should update caches as well, no?

i have the feeling that pypi internal arguments are dictating maintainer experience here... why not give maintainers the information (like end-user download stats) to make well informed decisions and let them make their decisions?

at least however put a warning on that delete button on the pypi website as mentioned in https://github.com/pypa/packaging-problems/issues/74#issuecomment-260407394 ... it's a big UX hazard that trips up maintainers and is luring in the interface for at least 1.5 years now.

**Comment by joelmiller** [comment]
When you go to remove something, you get a warning message that it can't be undone.  Could that warning be improved?  Not only can it not be undone, but you will not be able to upload a replacement with the same name...

**Comment by atagar** [comment]
Yes, that's what's missing from the message. "Cannot be undone" means "cannot recover the deleted file" but it says nothing about it being impossible to re-upload another file.

Ideally the message would say something like "This cannot be undone and another file by the same name cannot be uploaded. This is a known limitation of PyPi and tracked on [link to issue]."

**Comment by joernhees** [comment]
all that's missing is someone hitting a button here: https://github.com/pypa/pypi-legacy/pull/591

**Comment by nobodyinperson** [comment]
This is unbelievably annoying undocumented behaviour. Totally ruined my day.

**Comment by brainwane** [comment]
Some updates for folks in this discussion:

pypa/pypi-legacy#591 was merged in June 2017.

See https://github.com/pypa/packaging-problems/issues/75#issuecomment-347739479 for the ability to upload wheels for old releases, https://mail.python.org/pipermail/distutils-sig/2017-December/thread.html#31835 regarding immutability of metadata for past releases, and https://packaging.python.org/guides/using-testpypi/ for guidance on using the new Test PyPI.

Warehouse, the codebase behind the new PyPI, is available as a pre-production site at https://pypi.org, [we're making steady progress on the developer roadmap](https://wiki.python.org/psf/WarehouseRoadmap) thanks to [funding from Mozilla's Open Source Support Program](https://pyfound.blogspot.com/2017/11/the-psf-awarded-moss-grant-pypi.html), and it's on its way to replacing the legacy PyPI site entirely within the next few months. [We're currently seeking feedback from package maintainers](https://wiki.python.org/psf/WarehousePackageMaintainerTesting) about what does or doesn't work for you in the new interface, and in the next few weeks we'll be doing more outreach to the wider Python developer community. If you have tried the release deletion interface in Warehouse (including at https://test.pypi.org ) and the warning isn't strong enough, please let us know.

pypa/warehouse#726 is open to discuss how PyPI and Test PyPI can be better at providing staging/preview functionality for maintainers. #10 is open to discuss toolchain improvements for testing a package after creating it but before uploading it. There's an open issue pypa/twine#207 for adding a "dry run" option to [Twine](http://twine.readthedocs.io/). (I am a new maintainer of Twine and just released Twine 1.10.0; please upgrade.) @atagar I saw your comment https://github.com/pypa/packaging-problems/issues/74#issuecomment-261802260 and followed the [wiki history](https://trac.torproject.org/projects/tor/wiki/doc/stem?sfp_email=&sfph_mail=&action=diff&version=102&old_version=101&sfp_email=&sfph_mail= ) to check your code -- have you switched to Twine, and if so, does the approach you used still work?

Hope this helps.

**Comment by n1-cz** [comment]
Just ran into this. After reading almost whole thread I don't think there is something I can add. In once sentense - It's very stupid to have a test repository without the ability to do test stuff.

I hope the team will reconsider this. In the meantime we are going to spam your database with new and new "versions" because we can't rewrite the current one.

**Comment by sunbearc22** [comment]
I am using Test PyPI to try out my installation. I needed to make some changes to it but encountered errors. I finally tried deleting the entire package from Test PyPI, and made some changes to my script files, re-ran `python3 setup.py sdist bdist_wheel`and `twine upload --repository-url https://test.pypi.org/legacy/ dist/* --verbose`. Still had issue:

```
Uploading distributions to https://test.pypi.org/legacy/
Uploading upgrade_pip_packages-0.0.1.dev0-py3-none-any.whl
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7.32k/7.32k [00:01<00:00, 5.25kB/s]
Content received from server:
<html>
 <head>
  <title>400 File already exists. See https://test.pypi.org/help/#file-name-reuse</title>
 </head>
 <body>
  <h1>400 File already exists. See https://test.pypi.org/help/#file-name-reuse</h1>
  The server could not comply with the request since it is either malformed or otherwise incorrect.<br/><br/>
File already exists. See https://test.pypi.org/help/#file-name-reuse


 </body>
</html>
HTTPError: 400 Client Error: File already exists. See https://test.pypi.org/help/#file-name-reuse for url: https://test.pypi.org/legacy/
```

I had not seen this discussion thread before doing the above. After reading it, my questions are:
1. Am I correct to conclude that with Test PyPI, after deleting my package, I can no longer reuse the same package name even though I am the owner of the package name?
2. I thought Test PyPI is ephemeral in nature. If so, can I reuse the deleted package name a day later? 
3. According to [#file-name-reuse](https://test.pypi.org/help/#file-name-reuse), does the word "filename"  means package name or something else?

**Comment by benoit-pierre** [comment]
You cannot reuse the same distribution name, .i.e. reuse the same source distribution or wheel file name (even if the whole package was deleted). You need to change the version number (or increase the build number for wheels).

**Comment by sunbearc22** [comment]
@benoit-pierre Thanks for clarifying. I hope PyPI developers can make the Test PyPI interface host packages ephemerally (i.e. temporarily) as it is for testing purpose. New user like myself will make mistakes and discover issues after packaging and needs to make fixes before actually uploading the package into the PyPI interface. I like to request the distribution name used in this test environment be  non-permanent, reusable and not be linked to actual PyPI interface. My wish list.... Thank you.

**Comment by benoit-pierre** [comment]
You could use a release candidate version number: `foo-1.0rc1`, `foo-1.0rc2`, etc... Until all issues pushing to Test PyPi have been resolved and you release the final `foo-1.0` to PyPI.

**Comment by 1313e** [comment]
> You could use a release candidate version number: `foo-1.0rc1`, `foo-1.0rc2`, etc... Until all issues pushing to Test PyPi have been resolved and you release the final `foo-1.0` to PyPI.

Although I completely understand that, people often use the Test PyPI to "test" packaging, uploading and installing their packages.
Which often involves making very minimal changes to the code and reuploading it again, tens and sometimes hundreds of times.
It is incredibly annoying to have to remember yourself every time that you have to increase the version number after making a change.
I did the same thing once with a package of mine, and I ended up with version 'v1.0.0rc127' on Test PyPI, which is a bit too much in my opinion.

Unlike normal PyPI, Test PyPI should allow for the same version number to be uploaded multiple times, overwriting the current version if it already exists (as long as you own the package name).
As far as I know, Test PyPI is purely there for testing (hence the name) and therefore the policy that no distributions should ever be deleted (permanently) from the system, does not apply here.
I hope somebody can take a good look at this, since having to increase the version number for every test release on Test PyPI does not add anything beneficial to anybody besides being forced to learn to increase the version number for every release (which is required for PyPI).

**Comment by joelmiller** [comment]
>
> You could use a release candidate version number: foo-1.0rc1, foo-1.0rc2,
> etc... Until all issues pushing to Test PyPi have been resolved and you
> release the final foo-1.0 to PyPI.
>

Sure this can be done.

But the current warning does not advise people to take these steps.

A perfectly reasonable person can look at the instructions and expect that
"this [deletion] cannot be undone" does not mean "once you've deleted this,
you can't replace it with something else".

**Comment by pradyunsg** [comment]
> pypa/warehouse#726 is open to discuss how PyPI and Test PyPI can be better at providing staging/preview functionality for maintainers.

**Comment by aaaakshat** [comment]
> > If you could delete a package and replace it with something different but with the same version, it can break their program
> 
> That's not what I am asking for.
> 
> I am asking for putting back the package I removed.

Exactly, I'm dealing with the same issue. I wasn't aware that this was a restriction imposed upon filenames and I had a large revamp of my code structure and I chose to remove my project only to try and add it back and find this error.

**Comment by torfsen** [comment]
> FTR, I don't think there is any opposition to Test PyPI not allowing re-uploads, just nobody has done the work to make that possible.

@dstufft Could you give a brief description of what changes would be needed for this feature and how much effort it would be? This would help potential contributors (like me) to decide whether they can tackle it.

**Comment by alexlaurence** [comment]
You have to update your Github repo, then you can reupload.

**Comment by OneAdder** [comment]
I think it still looks pretty weird. I just had a problem with this: I accidentally built my package the wrong way and uploaded it. It was not a problem within the package: I just forgot to delete the old build.

Now I had to change the version name for PyPI and explicitly say in the description that versions 0.8.5 and 0.8.5.post0 are the same version

**Comment by pradyunsg** [comment]
You don't have to say that - post releases are meant to be the same as the existing release and only differ in metadata.

**Comment by MartinThoma** [comment]
One important thing seems not to be mentioned so far: **It is better for security and reliability to NOT allow changing the code of an uploaded version**. As a user of a 3rd party library, I can once check it and then pin the version. My program is guaranteed to work the same way. Even if the account of the maintainers get hacked, the worst thing that can happen is that the package is deleted. So it either works the same way or not at all. But at least the attacker cannot bring malicious code in another code base.

**Comment by Natim** [comment]
Note that if you remove the project alltogether and create it again, you can reupload code from previous version.

**Comment by 1313e** [comment]
> One important thing seems not to be mentioned so far: **It is better for security and reliability to NOT allow changing the code of an uploaded version**. As a user of a 3rd party library, I can once check it and then pin the version. My program is guaranteed to work the same way. Even if the account of the maintainers get hacked, the worst thing that can happen is that the package is deleted. So it either works the same way or not at all. But at least the attacker cannot bring malicious code in another code base.

Good point.
In that case, you could be allowed to delete and reupload a release if that happens within (let's say) a week after first release.
That way, we can replace releases when we made a mistake, while the problem you mentioned will not be an issue.

**Comment by brianmay** [comment]
Note that this has limitations. e.g. AFAIK I could upload a source file, and sometime later a malicious actor could upload a wheel file for the same version that has malicious content (i.e. it doesn't have to match the source).

**Comment by im-n1** [comment]
IMHO pypi.org should be immutable. test.pypi.org should be mutable because it's test environment.

**Comment by joelmiller** [comment]
I don't think many of the commenters have been suggesting that pypi itself should allow resubmissions with the same name, and I agree it should be immutable.  (but the error messages shouldn't say "deleting can't be undone" if in fact "deleted files can never be replaced").  There's a lot of problems there - different mirrors might have different versions, malicious actors, etc.

I'm okay with different mirrors of a test environment having different versions, and a malicious actor can't really get far by infecting something that's explicitly a test.

**Comment by ikamensh** [comment]
It is essential to disallow delete and upload of existing file. Without this you can't *guarantee* anything; malicious actor could inject code to your project and you wouldn't be able to do anything about it.

**Comment by brianmay** [comment]
@ikamensh It isn't quite that simple. I believe there is nothing stopping an upload of the same version of the package in another format, e.g. if you only upload a source file, a malicious wheel file can be uploaded at a later date, which the clients will use in preference. So I am not entirely convinced this is a security gain.  At least it does stop replacing already uploaded files however.

Strategies such as pipenv which store the checksum of the file are probably are the better approach to ensure the package doesn't change unexpectedly.

**Comment by brainwane** [comment]
I would welcome thoughts from @mnm678 and from @xmunoz on this.

**Comment by mnm678** [comment]
I agree with the general sentiment of this thread: once a package is uploaded to PyPI, it should not change. If an attacker was able to compromise a project, they should not be able to replace an existing package with arbitrary code. Not only that, but even a well intentioned change to an existing package version could break a client's code. If a client pins a version of a package, they should get the same code every time that package is downloaded.

Earlier in the thread, @dstufft suggested a soft-delete function which would allow developers to undo the deletion of a package without reuploading the file. This seems like a good compromise between the security risks of new uploads and the usability problem of accidental deletion. In this case an attacker could undo the deletion of a version, but would not be able to add malicious code to that version (although they could still add a new, bad version unless something like PEP 480 is used to verify the developer's identity).

However, the discussion about PyPI-test is a different story. The security risks in a test repository are less important as any packages are not used in production, so I don't see a problem with allowing test versions of packages to be replaced.

**Comment by uranusjr** [comment]
FYI the soft-deletion feature has been formalised as PEP 592 “yank.” It is not yet implemented for PyPI (pypa/warehouse#5837).

**Comment by brainwane** [comment]
Relatedly: an implementation proposal for a "draft releases" feature on PyPI https://github.com/pypa/warehouse/issues/726 is now being discussed at 
https://discuss.python.org/t/feature-proposal-for-pypi-draft-releases/3903 . Check it out to see whether you have comments to share there. @alanbato would like comments on the proposal by 30 April (10 days from now) so he can go ahead and start implementation work.

(More context [on distutils-sig](https://mail.python.org/archives/list/distutils-sig@python.org/message/G355FYOYQZ6HIH7RK47WPZOYHNCEXGOB/).)

**Comment by brainwane** [comment]
Now that [PEP 592 is accepted](https://www.python.org/dev/peps/pep-0592/) and implemented pypa/warehouse#5837, I hope people in this thread will consider looking at [the yanking feature](https://pypi.org/help/#yanked) and see which of your use cases it suits.

**Comment by TylerGubala** [comment]
> With the new policy you basically say: You've ONE SINGLE TRY and that one SHOULD WORK. No chance for a 2nd try. IMHO this isn't the purpose of a testing system and breaks the whole "we've a testing repo" idea. To be honest, I think this only leads to annoyed developers and a lot of "crippled versions" because developers couldn't properly test their versions before going live.

Literally this. In my repo, [Blenderpy](https://github.com/TylerGubala/blenderpy) I have the additional headache that I often don't know exactly everything to test in advance, as there can be a lot of edge cases in the precompiled C/C++ code that may be remedied by configuring with different `cmake` arguments.

> Why cannot you do `package x.y.z.dev0` and then `package x.y.z.dev1`?

For example, in my case people performing `pip install bpy==2.81` want the api version 2.81.

So now I effectively have my own version squatting on me, wherein people _factually cannot_ install the proper version because I made an honest mistake.

So I think there are a lot of important considerations here. There is the "version contract", but then there are case by case times where I think that does not apply. I should think that if I accidentally upload some bad version of my own repo, then go back and try to remedy the issue, that does not critically break the version contract.

I'm sorry but I do not understand the concept of shackling yourself to a "version contract" that basically screws yourself and your users in the end.

**Comment by arigo** [comment]
I was just told about "build numbers", which apparently let us upload several wheels for the same version number and platform, just by giving them slightly different names:

wheel-0.31.0-py2.py3-none-any.whl
wheel-0.31.0-1-py2.py3-none-any.whl
wheel-0.31.0-2-py2.py3-none-any.whl

It would seem to be exactly the feature that lets us re-upload files---the same file, or one with a fix.  It would have saved the OP and myself the pain of making completely new versions just to fix one obscure platform wheel---including, lately, *breaking* another, widely-used platforms while doing so.

Why is this not the solution?  Why is the "build number" feature not widely known?

**Comment by duck-rh** [comment]
I did not see any progress on #10 and related pypa/warehouse#726, so is there any hope to be able to reupload on testpypi so we can check the result before the final upload on pypi? I'm fine with another solution but it seems not much progress was made since such kind of discussions started in 2013.

**Comment by duck-rh** [comment]
@arigo I tried your trick to rename the files (thanks `rename`) in order to avoid changing the sources and ended-up with this error: "Only one sdist may be uploaded per release". Since I cannot exclude it to be sure my fixes are properly taken into account, I tried deleting the version and got the same error as this issue; fortunately bumping the "build number" again worked.

Seems the best workaround so far though.

**Comment by makseq** [comment]
@arigo Unfortunately I have a problem with this approach: 

```
pip install label-studio==0.9.0
DEPRECATION: Python 3.5 reached the end of its life on September 13th, 2020. Please upgrade your Python as Python 3.5 is no longer maintained. pip 21.0 will drop support for Python 3.5 in January 2021. pip 21.0 will remove support for this functionality.
Collecting label-studio==0.9.0
  Using cached label_studio-0.9.0-2-py3-none-any.whl (18.9 MB)
ERROR: Requested label-studio==0.9.0 from https://files.pythonhosted.org/packages/7f/32/da8e6b279134e96c8b835bba3c20b8638473fed4f514b10472582258ae8b/label_studio-0.9.0-2-py3-none-any.whl#sha256=1363a13887d6351b720c89c43be78d41d42fff1cee56cebdd4fbf60077668fb9 has different version in metadata: '0.9.0.post2'
```

**Comment by Tasty213** [comment]
Hi, I just removed a package version because my internet connection failed midway through and so it thought I had uploaded a file when I actually hadn't. I didn't realise this would mean I could never use that version number again because there's no warning. I think this is a bit absurd as now I'm going to have to make a new release (v0.1.4) without making any changes. 

There should be a warning before you delete versions and files saying that you can never use those names again to ensure users understand what they're doing.

**Comment by henryiii** [comment]
You can upload a post version, `v0.1.3.post1` - this will be treated like the original, `--upgrade` will know they are the same, etc. You can also upload missing files without starting a new version if one or more wheels didn't upload (or the SDist).

**Comment by microprediction** [comment]
I just bumped the version number a few times, tried again and it worked fine, and got on with my day.

**Comment by rzashakeri** [comment]
I also had the same problem and could not re-upload a package with the same version after deleting it from pypi, but the following solution worked for me.

https://stackoverflow.com/a/28874414/13175549

**Comment by octachrome** [comment]
If you append a build tag e.g. `-1` to the version number, this works and is considered by `pip` to be the same version. This behaviour is documented at https://peps.python.org/pep-0427/#file-name-convention:

> build tag
> Optional build number. Must start with a digit. Acts as a tie-breaker if two wheel file names are the same in all other respects (i.e. name, version, and other tags). Sort as an empty tuple if unspecified, else sort as a two-item tuple with the first item being the initial digits as an int, and the second item being the remainder of the tag as a str.

This worked for me with a binary package, e.g. I renamed `mypkg-1.2.3-cp310-cp310-win_amd64.whl` to `mypkg-1.2.3-1-cp310-cp310-win_amd64.whl`. It should work with source distributions too, but I didn't test that.

I initially tried appending `.post1` as mentioned in another comment, but that is considered a different version number by `pip` so users must change their `requirements.txt` etc., which was not acceptable for me.

**Comment by pfmoore** [comment]
> It should work with source distributions too, but I didn't test that.

Source distributions have no formal notion of a build tag, so if it works with sdists, that's likely to be by accident and shouldn't be relied on.
